Core data: harmonize getRevision selector and resolver function signatures

ramonjd opened the pull request:
## What?

Follow up for:

- https://github.com/WordPress/gutenberg/pull/54046

Yo!

This amazing PR:

1. Removes default value of query `{}` on `getRevision()`. It messes up selector memoization since the function signatures are different.
2. Resolves individual revisions in `getRevisions()`, the same way `getEntityRecords` does it.
3. Makes a minor grammatical upgrade.

## Why?
1. Selectors should match their resolvers! Also, calling `getRevisions()` then `getRevision()` doesn't return cached store item and replaces current items state.
2. Ensures that any subsequent calls to `getRevision` for revisions that we've already fetched via `getRevisions()` are marked as resolved and therefore do not trigger the resolver (and therefore an API call).
3. My English teacher was cruel.

## Testing Instructions
Check that the state survives multiple requests and that cached results are returned.

For example

### Before

```js
await wp.data.resolveSelect( 'core' ).getRevisions( 'postType', 'post', 1 ) // Returns full revisions collections

await wp.data.resolveSelect( 'core' ).getRevision( 'postType', 'post', 1, 30 ) // Returns single revision with API call.

await wp.data.resolveSelect( 'core' ).getRevisions( 'postType', 'post', 1 ) // Returns revisions collections with only a single revision
```

### After

```js
await wp.data.resolveSelect( 'core' ).getRevisions( 'postType', 'post', 1 ) // Returns full revisions collections

await wp.data.resolveSelect( 'core' ).getRevision( 'postType', 'post', 1, 30 ) // Returns single revision with NO API call.

await wp.data.resolveSelect( 'core' ).getRevisions( 'postType', 'post', 1 ) // Returns full revisions collections
```

ramonjd committed:
Remove default value of query {} on getRevision. It messes up selector memoization since the function signatures are different.

ramonjd committed:
Resolve individual revisions as per `getEntityRecords`. This ensures that any subsequent calls to `getRevision` for revisions that we've already fetched via `getRevisions()` are marked as resolved and therefore do not trigger the resolver (and therefore an API call).

andrewserong commented on packages/core-data/src/resolvers.js:
```diff
@@ -801,6 +801,26 @@ export const getRevisions =
 		);
+
+		// When requesting all fields, the list of results can be used to
+		// resolve the `getEntityRecord` selector in addition to `getEntityRecords`.
```
Looks like a copy / paste 😄

```suggestion
		// resolve the `getRevision` selector in addition to `getRevisions`.
```

andrewserong reviewed (commented):
Thanks for following up here! This appears to resolve the issue of calling `getRevision` resulting in an unnecessary API request if we already previously fetched the data as part of `getRevisions` 👍

An issue I ran into while testing, though, is that if I call `getRevision` to grab an individual global styles revision (prior to calling `getRevisions`), then the API call is still to fetch all revisions, rather than an individual one. I think the culprit might be that the following `getRevisionsUrl` signature doesn't support passing in a `revisionId`?

https://github.com/WordPress/gutenberg/blob/8eb63ddb51525164860e785900d7d86544f37afc/packages/core-data/src/entities.js#L214

Example call:

```
wp.data.select( 'core' ).getRevision( 'root', 'globalStyles', 4169, 4180 );
```

This winds up calling to get all revisions for `4169`:

<img width="660" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/d4f57f16-6d38-453f-b41c-9d7069538542">

Updating the `getRevisionsUrl` method to match what's used for `postType` seems to do the trick for me:

```
		getRevisionsUrl: ( parentId, revisionId ) =>
			`/wp/v2/global-styles/${ parentId }/revisions${
				revisionId ? '/' + revisionId : ''
			}`,
```

<img width="683" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/774f64b1-e79e-4b9b-a66d-f3a7f5de1a11">

Does that sound okay to you, or is the `root` `globalStyles` entity meant to behave differently to `postType`?

Apologies if I'm barking up the wrong tree!

ramonjd commented:
> An issue I ran into while testing, though, is that if I call getRevision to grab an individual global styles revision (prior to calling getRevisions), then the API call is still to fetch all revisions, rather than an individual one. I think the culprit might be that the following getRevisionsUrl signature doesn't support passing in a revisionId?

Ah yes!! Thanks for raising that. 

This change predates the following, recent update:

- https://github.com/WordPress/gutenberg/pull/55827

Now that we support revision URLs I think we're safe to copy it over as you say. 

> Apologies if I'm barking up the wrong tree!

You are too nice. This was a gross omission and deserves a proper flailing by the internet! 🤣

ramonjd committed:
Copy update.
Now that https://github.com/WordPress/gutenberg/pull/55827 has landed we can start supporting individual global styles URLs as well.

ramonjd committed:
While I'm here... :)

andrewserong reviewed (approved):
This is testing great for me now, @ramonjd, thanks!

For a moment I was worried that the `per_page` query param wasn't working for post revisions as I kept getting a response of only 10 revisions — then I double checked the post I was working on, and it happened to have only 10 revisions 🤦 😄

✅ `getRevisions` calls for posts and global styles works as on `trunk`
✅ `getRevision` calls for single post or single global style revisions will return the result from the resolved cache if found
✅ Using other query params still results in a fresh API call (the first time the fresh query params are used, with existing query params correctly matching a resolved response)

LGTM! ✨

ramonjd commented:
>  the per_page query param wasn't working for post revisions as I kept getting a response of only 10 revisions — then I double checked the post I was working on, and it happened to have only 10 revisions

Ha! Thanks a lot for testing so thoroughly.
